subsys/net/lib/dns/dispatcher.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(net_dns_dispatcher, CONFIG_DNS_SOCKET_DISPATCHER_LOG_LEVEL);
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/check.h>
+#include <zephyr/sys/slist.h>
+#include <zephyr/net/buf.h>
+#include <zephyr/net/dns_resolve.h>
+#include <zephyr/net/socket_service.h>
+
+#include "dns_pack.h"
+
+static K_MUTEX_DEFINE(lock);
+
+static sys_slist_t sockets;
+
+#define DNS_RESOLVER_MIN_BUF	1
+#define DNS_RESOLVER_BUF_CTR	(DNS_RESOLVER_MIN_BUF + \
+				 CONFIG_DNS_RESOLVER_ADDITIONAL_BUF_CTR)
+
+NET_BUF_POOL_DEFINE(dns_msg_pool, DNS_RESOLVER_BUF_CTR,
+		    DNS_RESOLVER_MAX_BUF_SIZE, 0, NULL);
+
+static int dns_dispatch(struct dns_socket_dispatcher *dispatcher,
+			int sock, struct sockaddr *addr, size_t addrlen,
+			struct net_buf *dns_data, size_t buf_len)
+{
+	/* Helper struct to track the dns msg received from the server */
+	struct dns_msg_t dns_msg;
+	bool is_query;
+	int data_len;
+	int ret;
+
+	data_len = MIN(buf_len, DNS_RESOLVER_MAX_BUF_SIZE);
+
+	dns_msg.msg = dns_data->data;
+	dns_msg.msg_size = data_len;
+
+	/* Make sure that we can read DNS id, flags and rcode */
+	if (dns_msg.msg_size < (sizeof(uint16_t) + sizeof(uint16_t))) {
+		ret = -EINVAL;
+		goto done;
+	}
+
+	if (dns_header_rcode(dns_msg.msg) == DNS_HEADER_REFUSED) {
+		ret = -EINVAL;
+		goto done;
+	}
+
+	is_query = (dns_header_qr(dns_msg.msg) == DNS_QUERY);
+	if (is_query) {
+		if (dispatcher->type == DNS_SOCKET_RESPONDER) {
+			/* Call the responder callback */
+			ret = dispatcher->cb(dispatcher->ctx, sock,
+					     addr, addrlen,
+					     dns_data, data_len);
+		} else if (dispatcher->pair) {
+			if (dispatcher->pair->type == DNS_SOCKET_RESPONDER) {
+				ret = dispatcher->pair->cb(dispatcher->pair->ctx, sock,
+							   addr, addrlen,
+							   dns_data, data_len);
+			} else {
+				/* The pair must be a resolver */
+				CHECKIF(dispatcher->pair->type != DNS_SOCKET_RESOLVER) {
```

Comment:
Perhaps we should also check for errors like this in `dns_dispatcher_register()`? So that you don't need the actual communication going on to figure out something is wrong.

Reply:
Indeed, good point, will change this.

Reply:
Actually we already check in `dns_dispatcher_register()` that the dispatcher type is not the same, so this checkif is not really needed.